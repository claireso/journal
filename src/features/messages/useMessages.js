import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

const MessagesContext = React.createContext()

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

export const MessagesProvider = ({ children }) => {
  const [state, setState] = useState([])

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = () => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status) => (message) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
