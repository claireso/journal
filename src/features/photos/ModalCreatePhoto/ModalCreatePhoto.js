import { memo } from 'react'
import PropTypes from 'prop-types'

import { Heading2 } from '@components/Headings'
import FormPhoto from '../FormPhoto'

const ModalCreatePhoto = ({ onSubmit, isProcessing }) => {
  return (
    <>
      <Heading2>Create a photo</Heading2>
      <FormPhoto onSubmit={onSubmit} isProcessing={isProcessing} />
    </>
  )
}

ModalCreatePhoto.propTypes = {
  onSubmit: PropTypes.func,
  isProcessing: PropTypes.bool
}

ModalCreatePhoto.defaultProps = {
  onSubmit: () => {},
  isProcessing: false
}

export default memo(ModalCreatePhoto)
