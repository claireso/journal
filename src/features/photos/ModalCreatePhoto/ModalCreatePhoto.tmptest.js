//@TODO restore test

// import { render, fireEvent, act } from '@testing-library/react'

// import ModalCreatePhoto from './ModalCreatePhoto'

// // import PhotosReducer from '../../reducer'
// import { PhotosProvider } from '../usePhotos'
// import { MessagesProvider } from '../../messages/useMessages'

// // const { PhotosProvider, INITIAL_STATE } = PhotosReducer

// // const { usePhotosReducer: originalUsePhotosReducer } = PhotosReducer

// describe('<Create />', () => {
//   const renderComponent = () => {
//     return render(
//       <MessagesProvider>
//         <PhotosProvider>
//           <ModalCreatePhoto />
//         </PhotosProvider>
//       </MessagesProvider>
//     )
//   }

//   // afterEach(() => {
//   //   PhotosReducer.usePhotosReducer = originalUsePhotosReducer
//   // })

//   test('should render component', () => {
//     const { container } = renderComponent()

//     expect(container.querySelector('input[name="color"]')).toBeNull()
//     expect(container).toMatchSnapshot()
//   })

//   test('should submit form', async () => {
//     const spySubmit = jest.fn()
//     // PhotosReducer.usePhotosReducer = () => [INITIAL_STATE, { createResource: spySubmit }]

//     const { container, getByLabelText } = renderComponent()

//     fireEvent.change(getByLabelText(/title/i), {
//       target: { value: 'Photo title' }
//     })
//     fireEvent.change(getByLabelText(/description/i), {
//       target: { value: 'Photo description' }
//     })

//     const inputFile = container.querySelector('input[type="file"]')

//     await act(async () => {
//       await fireEvent.change(inputFile, {
//         target: {
//           files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
//         }
//       })
//     })

//     fireEvent.change(getByLabelText(/position/i), {
//       target: { value: 'center' }
//     })

//     fireEvent.click(container.querySelector('input[type="submit"]'))

//     const expectedResult = new FormData(container.querySelector('form'))

//     expect(spySubmit).toHaveBeenCalledWith(expectedResult)
//   })

//   // test('should not submit form', async () => {
//   //   PhotosReducer.usePhotosReducer = () => [{ status: 'pending' }, {}]

//   //   const { container } = renderComponent()

//   //   expect(container.querySelector('input[type="submit"]')).toBeNull()
//   // })
// })
