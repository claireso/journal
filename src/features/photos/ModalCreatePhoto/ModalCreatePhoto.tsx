import { memo } from 'react'

import { Heading2 } from '@components/Headings'
import FormPhoto from '../FormPhoto'

interface ModalCreatePhotoProps {
  onSubmit: (data: FormData) => void
  isProcessing?: boolean
}

const ModalCreatePhoto = ({ onSubmit, isProcessing = false }: ModalCreatePhotoProps) => {
  return (
    <>
      <Heading2>Create a photo</Heading2>
      <FormPhoto onSubmit={onSubmit} isProcessing={isProcessing} />
    </>
  )
}

export default memo(ModalCreatePhoto)
