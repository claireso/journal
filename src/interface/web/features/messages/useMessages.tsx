'use client'

import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

interface Message {
  key?: string
  status: 'default' | 'success' | 'error' | 'info'
  message: string
}

interface Actions {
  displaySuccessMessage: (data: Omit<Message, 'status'>) => void
  displayErrorMessage: (data: Omit<Message, 'status'>) => void
  closeMessage: (index: number) => void
  closeAllMessages: () => void
}

const INITIAL_STATE: Message[] = []

type TMessagesContext<T> = [T, React.Dispatch<React.SetStateAction<T>>]

const MessagesContext = React.createContext<TMessagesContext<Message[]>>([null!, null!])

interface MessagesProviderProps {
  children: React.ReactNode
}

export const MessagesProvider = ({ children }: MessagesProviderProps) => {
  const [state, setState] = useState(INITIAL_STATE)

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = (): [Message[], Actions] => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index: number) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status: Message['status']) => (message: Omit<Message, 'status'>) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message: Message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
