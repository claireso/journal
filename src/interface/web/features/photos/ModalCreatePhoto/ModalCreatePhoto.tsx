import { memo } from 'react'

import { createPhoto } from '@application/usecases'
import AdminModal from '@web/features/modal/AdminModal'
import FormPhoto from '../FormPhoto'

const ModalCreatePhoto = () => {
  return (
    <AdminModal title="Create photo">
      <FormPhoto
        action={createPhoto}
        successMessage={{ key: 'CRUD_PHOTO', message: 'Your photo has been created successfully' }}
        errorMessage={{ key: 'CRUD_PHOTO', message: 'An error has occured during the creation. Please retry' }}
      />
    </AdminModal>
  )
}

export default memo(ModalCreatePhoto)
