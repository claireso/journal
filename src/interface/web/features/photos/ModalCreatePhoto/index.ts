import ModalCreatePhoto from './ModalCreatePhoto'

export default ModalCreatePhoto
