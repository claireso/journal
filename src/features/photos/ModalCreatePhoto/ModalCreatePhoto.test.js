import { screen, render, fireEvent, act } from '@testing-library/react'

import ModalCreatePhoto from './ModalCreatePhoto'

import { MessagesProvider } from '../../messages/useMessages'

describe('<ModalCreatePhoto />', () => {
  const renderComponent = (props = {}) => {
    return render(
      <MessagesProvider>
        <ModalCreatePhoto {...props} />
      </MessagesProvider>
    )
  }

  it('should render component', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should create photo', async () => {
    const props = {
      onSubmit: jest.fn()
    }

    renderComponent(props)

    fireEvent.change(screen.getByLabelText(/title/i), {
      target: { value: 'Photo title' }
    })
    fireEvent.change(screen.getByLabelText(/description/i), {
      target: { value: 'Photo description' }
    })

    const inputFile = document.querySelector('input[type="file"]')

    await act(async () => {
      await fireEvent.change(inputFile, {
        target: {
          files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
        }
      })
    })

    fireEvent.change(screen.getByLabelText(/position/i), {
      target: { value: 'center' }
    })

    fireEvent.click(screen.getByTestId('submit'))

    const formData = new FormData(document.querySelector('form'))

    expect(props.onSubmit).toHaveBeenCalledWith(formData)
  })
})
