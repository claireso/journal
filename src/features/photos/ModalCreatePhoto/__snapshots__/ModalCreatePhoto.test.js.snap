// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalCreatePhoto /> should render component 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 2.2rem;
  margin: 0 0 3rem;
  font-weight: 500;
}

<h2
    class="c0"
  >
    Create a photo
  </h2>
  .c0 {
  margin: 0 0 var(--gutter);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 > label {
  margin: 0 1rem 0 0;
}

.c1 {
  display: block;
  font-size: 1.6rem;
  line-height: 1.5;
  margin: 0 0 1.2rem;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--form-input-bg);
  border: 1px solid var(--border-color);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  display: block;
  font-family: var(--font-family);
  outline: none;
  padding: 0.8rem 1.2rem;
  width: 100%;
  font-size: var(--font-size-normal);
  line-height: 1.71;
  -webkit-transition: border-color 250ms ease-out;
  transition: border-color 250ms ease-out;
}

.c2:focus {
  border-color: var(--border-color-focus);
}

.c7 {
  position: relative;
}

.c7 svg {
  position: absolute;
  top: 50%;
  right: 0.9rem;
  bottom: 0;
  -webkit-transform: translate(0,-50%) rotate(90deg);
  -ms-transform: translate(0,-50%) rotate(90deg);
  transform: translate(0,-50%) rotate(90deg);
  pointer-events: none;
}

.c8 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--form-input-bg);
  border: 1px solid var(--border-color);
  border-radius: 0.5rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  cursor: pointer;
  font-family: 'Roboto',Arial,sans-serif;
  font-size: var(--font-size-normal);
  outline: none;
  padding: 1.1rem 2.6rem 1.1rem 1.1rem;
}

.c8:focus {
  border-color: var(--border-color-focus);
}

.c3 {
  border: 1px solid var(--border-color);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  overflow: hidden;
  position: relative;
}

.c4 {
  padding: 2rem;
}

.c4 svg {
  display: block;
  width: 2.4rem;
  height: 2.4rem;
  fill: var(--primary-normal);
  margin: 0 auto;
}

.c4 span {
  display: block;
  font-size: var(--font-size-normal);
  margin: 1rem 0 0;
  text-align: center;
}

.c4 small {
  color: var(--border-color-focus);
  display: block;
  font-size: var(--font-size-smaller);
  margin: 0.5rem 0 0;
  text-align: center;
}

.c5 {
  bottom: 0;
  cursor: pointer;
  left: 0;
  opacity: 0;
  position: absolute;
  right: 0;
  top: 0;
  width: 100%;
}

.c9 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--primary-normal);
  border: none;
  border-radius: 3.6rem;
  color: white;
  cursor: pointer;
  display: block;
  font-size: var(--font-size-normal);
  line-height: 1;
  margin: 2.8rem auto 0;
  outline: none;
  padding: 1.6rem 5.2rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c9:hover {
  background: var(--primary-lighter);
}

<form
    action=""
    enctype="multipart/form-data"
    method="POST"
  >
    <div
      class="c0"
    >
      <label
        class="c1"
        for="title"
      >
        Title
      </label>
      <input
        class="c2"
        data-testid="title"
        id="title"
        name="title"
        type="text"
        value=""
      />
    </div>
    <div
      class="c0"
    >
      <label
        class="c1"
        for="description"
      >
        Description
      </label>
      <input
        class="c2"
        data-testid="description"
        id="description"
        name="description"
        type="text"
        value=""
      />
    </div>
    <div
      class="c0"
    >
      <label
        class="c1"
        for="file"
      >
        Photo
      </label>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <svg
            height="16"
            viewBox="0 0 26 28"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 23c0-0.547-0.453-1-1-1s-1 0.453-1 1 0.453 1 1 1 1-0.453 1-1zM24 23c0-0.547-0.453-1-1-1s-1 0.453-1 1 0.453 1 1 1 1-0.453 1-1zM26 19.5v5c0 0.828-0.672 1.5-1.5 1.5h-23c-0.828 0-1.5-0.672-1.5-1.5v-5c0-0.828 0.672-1.5 1.5-1.5h6.672c0.422 1.156 1.531 2 2.828 2h4c1.297 0 2.406-0.844 2.828-2h6.672c0.828 0 1.5 0.672 1.5 1.5zM20.922 9.375c-0.156 0.375-0.516 0.625-0.922 0.625h-4v7c0 0.547-0.453 1-1 1h-4c-0.547 0-1-0.453-1-1v-7h-4c-0.406 0-0.766-0.25-0.922-0.625-0.156-0.359-0.078-0.797 0.219-1.078l7-7c0.187-0.203 0.453-0.297 0.703-0.297s0.516 0.094 0.703 0.297l7 7c0.297 0.281 0.375 0.719 0.219 1.078z"
            />
          </svg>
          <span>
            Upload new photo
          </span>
          <small>
            (only jpg and png)
          </small>
          <input
            class="c5"
            name="file"
            required=""
            type="file"
          />
        </div>
      </div>
    </div>
    <div
      class="c0 c6"
    >
      <label
        class="c1"
        for="position"
      >
        Position
      </label>
      <div
        class="c7"
      >
        <select
          class="c8"
          id="position"
          name="position"
          required=""
        >
          <option
            selected=""
            value="left"
          >
            Left
          </option>
          <option
            value="center"
          >
            Center
          </option>
          <option
            value="right"
          >
            Right
          </option>
        </select>
        <svg
          height="14"
          viewBox="0 0 9 28"
          width="9"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.297 15c0 0.125-0.063 0.266-0.156 0.359l-7.281 7.281c-0.094 0.094-0.234 0.156-0.359 0.156s-0.266-0.063-0.359-0.156l-0.781-0.781c-0.094-0.094-0.156-0.219-0.156-0.359 0-0.125 0.063-0.266 0.156-0.359l6.141-6.141-6.141-6.141c-0.094-0.094-0.156-0.234-0.156-0.359s0.063-0.266 0.156-0.359l0.781-0.781c0.094-0.094 0.234-0.156 0.359-0.156s0.266 0.063 0.359 0.156l7.281 7.281c0.094 0.094 0.156 0.234 0.156 0.359z"
          />
        </svg>
      </div>
    </div>
    <input
      class="c9"
      data-testid="submit"
      type="submit"
      value="Create"
    />
  </form>
</DocumentFragment>
`;
